File: go/authmiddleware/revocation.go
package authmiddleware

import (
	"context"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"
)

type RevocationChecker struct {
	rdb       *redis.Client
	keyPrefix string

	mu    sync.RWMutex
	cache map[string]time.Time
	ttl   time.Duration
}

func NewRevocationChecker(rdb *redis.Client) *RevocationChecker {
	return &RevocationChecker{
		rdb:       rdb,
		keyPrefix: "revoked:",
		cache:     make(map[string]time.Time),
		ttl:       60 * time.Second,
	}
}

func (r *RevocationChecker) IsRevoked(ctx context.Context, jti string) (bool, error) {
	if jti == "" {
		return false, nil
	}

	r.mu.RLock()
	if exp, ok := r.cache[jti]; ok && time.Now().Before(exp) {
		r.mu.RUnlock()
		return true, nil
	}
	r.mu.RUnlock()

	n, err := r.rdb.Exists(ctx, r.keyPrefix+jti).Result()
	if err != nil {
		return false, err
	}
	revoked := n > 0
	if revoked {
		r.mu.Lock()
		r.cache[jti] = time.Now().Add(r.ttl)
		r.mu.Unlock()
	}
	return revoked, nil
}

File: rust/auth-middleware/src/revocation.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};

use redis::AsyncCommands;
use tokio::sync::RwLock;

pub struct RevocationChecker {
    client: redis::Client,
    cache: RwLock<HashMap<String, Instant>>,
    cache_ttl: Duration,
    key_prefix: String,
}

impl RevocationChecker {
    pub fn new(client: redis::Client) -> Arc<Self> {
        Arc::new(Self {
            client,
            cache: RwLock::new(HashMap::new()),
            cache_ttl: Duration::from_secs(60),
            key_prefix: "revoked:".to_string(),
        })
    }

    pub async fn is_revoked(&self, jti: &str) -> Result<bool, redis::RedisError> {
        if jti.is_empty() {
            return Ok(false);
        }

        {
            let guard = self.cache.read().await;
            if let Some(exp) = guard.get(jti) {
                if *exp > Instant::now() {
                    return Ok(true);
                }
            }
        }

        let mut conn = self.client.get_multiplexed_async_connection().await?;
        let key = format!("{}{}", self.key_prefix, jti);
        let exists: bool = conn.exists(&key).await?;

        if exists {
            let mut guard = self.cache.write().await;
            guard.insert(jti.to_string(), Instant::now() + self.cache_ttl);
        }
        Ok(exists)
    }
}
